### PR TITLE
[AIR][Doc] New Example: LightningTrainer with experiment tracking tools

### LINKED ISSUE #34869: [Doc][Train] Add Experiment Tracking Tools with LightningTrainer Example
### Description

Some users have trouble using wandb/cometML for experiment tracking while onboarding LightningTrainer. Post a new example to demo the best practices.

ptl-wandb-results
ptl-comet-results
ptl-tensroboard-results
ptl-mlflow-results

### Link

_No response_

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

Some users have trouble using wandb/cometML for experiment tracking while onboarding LightningTrainer. They are struggling between PTL native logger and our Ray integration.

This example aims to show the best practices of how to use PTL native logger in LightningTrainer.

Rendered doc: https://anyscale-ray--34812.com.readthedocs.build/en/34812/train/examples/lightning/lightning_exp_tracking.html

## Related issue number

Close #34869
## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+    "\n",
+    "    # Wandb\n",
+    "    wandb_api_key = os.environ.get(\"WANDB_API_KEY\", None)\n",
+    "    wandb_logger = WandbLogger(\n",
+    "        name=name, \n",
+    "        project=project_name, \n",
+    "        save_dir=f\"{save_dir}/wandb\", \n",
+    "        offline=offline\n",
+    "    )\n",
+    "    callbacks = [] if offline else [WandbLoginCallback(key=wandb_api_key)]\n",
```
Comment (justinvyu) [inline doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:137]: Hmm, interesting. Is this the recommended way of doing wandb auth? What if we just make the API key env var set on all workers?

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+    "loggers, callbacks = create_loggers(\n",
+    "    name=\"demo-run\", project_name=\"demo-project\", offline=True\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Train the model and check out your logs:"
```
Comment (justinvyu) [inline doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:196]:
```suggestion
    "## Train the model and view logged results"
```

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+     "data": {
+      "text/html": [
+       "<div class=\"tuneStatus\">\n",
+       "  <div style=\"display: flex;flex-direction: row\">\n",
+       "    <div style=\"display: flex;flex-direction: column;\">\n",
+       "      <h3>Tune Status</h3>\n",
+       "      <table>\n",
+       "<tbody>\n",
+       "<tr><td>Current time:</td><td>2023-04-28 12:46:53</td></tr>\n",
+       "<tr><td>Running for: </td><td>00:01:20.41        </td></tr>\n",
```
Comment (justinvyu) [inline doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:214]: We can clear the output from this last cell -- we already see the outputs from the images.

**Review by justinvyu — COMMENTED** [review]
Nice! Can we also add this example to the table of contents under the other lightning examples.

Nit: Also, I think all lightning examples are tested on GPU right now, but this example only needs CPU? Maybe we can split the folder into `examples/lightning/cpu` and `examples/lightning/gpu`.

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "(lightning_experiment_tracking)=\n",
+    "\n",
+    "# Using W&B, CometML, and MLFlow Logger in LightningTrainer\n",
+    "\n",
+    "W&B, CometML, and MLFlow are all popular tools in the field of machine learning for managing, visualizing, and tracking experiments. \n",
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:12]: Make sense!

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb (hunk 1 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:137]: Yeah if we already set the API key, then we don't have to create this callback.

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define your loggers\n",
+    "\n",
+    "For offline loggers, no changes are required in the Logger initialization.\n",
+    "\n",
+    "For online loggers (W&B and CometML), you need to do two things:\n",
+    "- Set up your API keys as environment variables.\n",
+    "- Set `rank_zero_only.rank = None` to avoid Lightning creating a new experiment run on the driver node. "
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:91]: Nope. This only happens for Ray. Vanilla pytorch lightning script doesn't have driver process, the initialization of Logger runs on rank_0 worker directly.

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "(lightning_experiment_tracking)=\n",
+    "\n",
+    "# Using W&B, CometML, and MLFlow Logger in LightningTrainer\n",
+    "\n",
+    "W&B, CometML, and MLFlow are all popular tools in the field of machine learning for managing, visualizing, and tracking experiments. \n",
+    "\n",
+    "PyTorch Lightning provides built-in support for logging metrics to W&B and CometML during training, allowing users to easily track their model performance over time. MLFlow can also be used to track experiments and store artifacts such as model hyperparameters. With AIR LightningTrainer, you can still use these powerful tools for experiment tracking with the PyTorch Lightning's `logger`.\n"
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_exp_tracking.ipynb]: Thx! Done

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb (hunk 3 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_exp_tracking.ipynb]: Done!

**Comment by woshiyyya** [comment]
> Maybe we can split the folder into examples/lightning/cpu and examples/lightning/gpu.

@justinvyu Created a separate build object that test this example with CPU only.

**Review by justinvyu — APPROVED** [review]
LGTM!

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+   "source": [
+    "(lightning_experiment_tracking)=\n",
+    "\n",
+    "# Using Experiment Tracking Tools in LightningTrainer\n",
+    "\n",
+    "W&B, CometML, MLFlow, and Tensorboard are all popular tools in the field of machine learning for managing, visualizing, and tracking experiments. The {class}`~ray.train.lightning.LightningTrainer` integration in Ray AIR allows you to continue using these built-in experiment tracking integrations.\n",
+    "\n",
+    "\n",
+    ":::{note}\n",
+    "This guide shows how to use the native [Logger](https://lightning.ai/docs/pytorch/stable/extensions/logging.html) integrations in PyTorch Lightning. Ray AIR also provides {ref}`experiment tracking integrations <tune-exp-tracking-ref>` for all the tools mentioned in this example. We recommend sticking with the PyTorch Lightning loggers.\n",
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:16]: Is there any scenario the user would want to use the AIR integrations instead?

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define your model and dataloader\n",
+    "\n",
+    "No need for any code change here. We simply create a dummy model with dummy datasets for demo. We reported 3 metrics(\"train_loss\", \"metric_1\", \"metric_2\") in the training loop. Lightning's `Loggers` will capture and report them to the corresponding experiment tracking tools."
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:27]:
```suggestion
    "In this example, we simply create a dummy model with dummy datasets for demonstration. There is no need for any code change here. We report 3 metrics(\"train_loss\", \"metric_1\", \"metric_2\") in the training loop. Lightning's `Logger`s will capture and report them to the corresponding experiment tracking tools."
```

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    "    name=\"demo-run\", project_name=\"demo-project\", offline=True\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Train the model and view logged results:"
```
Comment (matthewdeng) [inline doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:200]:
```suggestion
    "## Train the model and view logged results"
```

**Review by matthewdeng — APPROVED** [review]
Looks great! Few minor nits

File: doc/source/train/examples/lightning/lightning_exp_tracking.ipynb (hunk 7 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_exp_tracking.ipynb:16]: Yes. One case I can think of is when a user wants more flexible logging, our integration allows users to directly interact with (like wandb) running objects. And Lightning's logger has another layer of abstraction, so some APIs may not be supported by Lightning's logger.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**